File: laboratorio9/Excepciones.py
class DepositoInvalidoException(Exception):
    def __init__(self, mensaje):
        super().__init__(mensaje)

class RetiroInvalidoException(Exception):
    def __init__(self, mensaje):
        super().__init__(mensaje)

class FondosInsuficientesException(Exception):
    def __init__(self, mensaje):
        super().__init__(mensaje)

class CuentaBancaria:
    def __init__(self, numero_cuenta, saldo):
        self.numero_cuenta = numero_cuenta
        self.saldo = saldo

    def depositar(self, n_saldo):
        try:
            if n_saldo < 0:
                raise DepositoInvalidoException("No se puede depositar una cantidad negativa")
            self.saldo += n_saldo
            print(f"Depósito exitoso. Saldo total: {self.saldo}")
        except DepositoInvalidoException as e:
            print(f"Error: {e}")

    def retirar(self, n_saldo):
        try:
            if n_saldo < 0:
                raise RetiroInvalidoException("No se puede retirar una cantidad negativa")
            if n_saldo > self.saldo:
                raise FondosInsuficientesException("No se puede retirar una cantidad mayor a la de su saldo")
            self.saldo -= n_saldo
            print(f"Retiro exitoso. Saldo restante: {self.saldo}")
        except (RetiroInvalidoException, FondosInsuficientesException) as e:
            print(f"Error: {e}")

if __name__ == "__main__":
    cuenta = CuentaBancaria("CWXS", 500.0)

    while True:
        try:
            print("\n--- Menú ---")
            print(f"Saldo actual: ${cuenta.saldo}")
            print("1.- Depósito")
            print("2.- Retiro")
            print("3.- Salir")

            opcion = int(input("Ingresa una opción: "))

            if opcion == 1:
                monto = float(input("Ingresa el monto a depositar: "))
                cuenta.depositar(monto)

            elif opcion == 2:
                monto = float(input("Ingresa el monto a retirar: "))
                cuenta.retirar(monto)

            elif opcion == 3:
                print("Saliendo del programa...")
                break

            else:
                print("Escoge una opción dentro de lo requerido")
        except ValueError:
            print("Error: Entrada inválida. Por favor, ingresa un número.")

File: laboratorio9/Excepciones.java
package laboratorio9;

import java.util.Scanner;

class DepositoInvalidoException extends Exception{
    public DepositoInvalidoException(String mensaje){
        super(mensaje);
    }
}

class RetiroInvalidoException extends Exception{
    public RetiroInvalidoException(String mensaje){
        super(mensaje);
    }
}

class FondosInsuficientesException extends Exception{
    public FondosInsuficientesException(String mensaje){
        super(mensaje);
    }
}

class CuentaBancaria{
    protected String numeroCuenta;
    protected double saldo;

    public CuentaBancaria(String numeroCuenta, double saldo){
        this.numeroCuenta = numeroCuenta;
        this.saldo = saldo;
    }

    public void Depositar(double n_saldo){
        try {
            if (n_saldo<0) {
                throw new DepositoInvalidoException("No se puede depositar una cantidad negativa");
            }
            this.saldo += n_saldo;
            System.out.println("Deposito exitoso. Saldo total: " + saldo);
        } catch (DepositoInvalidoException e) {
            System.out.println("Error: " + e.getMessage());
        }
    }

    public void Retirar(double n_saldo){
        try {
            if (n_saldo<0) {
                throw new RetiroInvalidoException("No se puede retirar una cantidad negativa");
            }
            if (n_saldo > saldo) {
                throw new FondosInsuficientesException("No se puede retirar una cantidad mayor a la de su saldo");
            }
            saldo -= n_saldo;
            System.out.println("Retiro exitoso. Saldo restante: " + saldo);
        } catch (RetiroInvalidoException | FondosInsuficientesException e) {
            System.out.println("Error: " + e.getMessage());
        }
    }
}

public class Excepciones{
    public static void main(String[] args) {
        CuentaBancaria cuenta = new CuentaBancaria("CWXS", 500.0);

        try(Scanner myObj = new Scanner(System.in)){
            int Option=0;
            double monto;
            do {
                try {
                    System.out.println("\n--- Menu ---");
                    System.out.println("Saldo actual: $" + cuenta.saldo);
                    System.out.println("1.- Deposito");
                    System.out.println("2.- Retiro");
                    System.out.println("3.- Exit");

                    System.out.println("Ingresa una opcion: ");
                    Option = myObj.nextInt();

                    switch (Option) {
                        case 1 -> {
                            System.out.println("Ingresa el monto a depositar:");
                            monto = myObj.nextDouble();

                            cuenta.Depositar(monto);
                        }

                        case 2 -> {
                            System.out.println("Ingresa el monto a retirar:");
                            monto = myObj.nextDouble();

                            cuenta.Retirar(monto);
                        }

                        case 3 -> {
                            System.out.println("Saliendo del programa...");
                            break;
                        }
                    
                        default -> System.out.println("Escoge una opcion dentro de lo requerido");
                    }
                } catch (Exception e) {
                    System.out.println("Error: Entrada inválida. Por favor, ingresa un número.");
                    myObj.nextLine(); // Limpiar el buffer
                }
            } while(Option != 3);
        }        
    }
}